### PR TITLE
Solr Exporters

### PR DESCRIPTION
**Pull request by mbarnett**
Building on the last two PRs, this Pull Request takes Solrizer out of the loop for indexing ActiveFedora objects for Solr.

A new type of helper model, "Exporters", is introduced, along with an implementation of a simple Solr Exporter DSL and Solr Exporters for our existing ActiveFedora-based models (Item, Thesis, Collection, Community, and FileSet). These Solr Exporters, via the DSL implementation in `Exporters::Solr::BaseExporter`, assuming the serialization-to-solr duties previously provided by the Solrizer gem. The search infrastructure and LockedLDPObject ActiveFedora wrapper are altered to use this new provider of Solr mapping.

Additionally, certain low-level Solr interface functionality is re-organized into a new `models/jupiter_core/solr_services` namespace for better clarity and isolation.

### CODE REVIEW

**Comment by mbarnett** [comment]
Apologies for the length of the PR, but it became kind of an all-or-nothing shift with the exporters to keep the tests passing without adding a bunch of workarounds to let Solr generation work both ways for intermediate PRs.

A *lot* of the length is just moving information about how attributes are to be represented for Solr from the models into separate exporters, so it's not quite as bad as it looks.

File: test/models/jupiter_core/locked_ldp_object_test.rb [hunk 1]
```diff
@@ -375,38 +391,15 @@ def unlocked_method_dont_let_locked_methods_mutate(attempted_title)
     end
   end
 
-  # You might ask yourself, "Should we be writing tests to validate basic functionality of upstream dependencies?"
-  #
-  # No.
-  # We should not _have_ to.
-  #
-  # (ﾉ °益°)ﾉ 彡 ┻━┻
-  test 'we are using a branch of ActiveFedora and Solrizer where index type works' do
```
Comment (mbarnett) [inline test/models/jupiter_core/locked_ldp_object_test.rb]: This test I was able to remove entirely, because it specifically tests a monkeypatch for a fix to ActiveFedora-Solrizer bugs that we no longer exposed to because we no longer declare indexes through ActiveFedora to be handled by Solrizer

File: app/models/jupiter_core/locked_ldp_object.rb [hunk 2]
```diff
@@ -412,6 +416,29 @@ def to_partial_path
...
+
+  # The solr exporter that has been declared for a given class. nil if one has not been declared
+  # OR if the declaration hasn't been run yet (as in early class-inheritance hooks, see the note on has_attribute's
+  # implementation)
+  def self.solr_exporter_class
+    @solr_exporter_class
+  end
+
+  # allows access to a solr exporter's map of mangled solr index names to the original attribute names through
+  # the class. The usecase for this is largely faceted searching, which receives facet results in terms of the
```
Comment (pgwillia) [inline app/models/jupiter_core/locked_ldp_object.rb]: Will also be useful for Department/Supervisor searches where the query needs to have a mangled Solr name but the user doesn't need to know that.

File: app/models/jupiter_core/solr_services/name_mangling.rb [hunk 3]
```diff
@@ -8,15 +8,15 @@ module JupiterCore::SolrNameMangler
     },
     integer: {
       search: 'isim',
-      sort: 'isi',
-      range_facet: 'isi',
+      sort: 'isi', # singluar
```
Comment (pgwillia) [inline app/models/jupiter_core/solr_services/name_mangling.rb]: Who are the `singular` comments for?  Why is this important?

The `SOLR_MANGLED_STEMS_BY_TYPE` is the part of me that feels a little awkward.  There is a coding to the dynamic field types -- type, stored, indexed, multivalued, etc.  So you kinda have to know this to know which is appropriate for its role https://lucene.apache.org/solr/guide/6_6/field-properties-by-use-case.html  I like that the Name Mangler makes the type and role very clear to the consumer, but this could also be done by improving the dynamic type naming on the Solr side.

File: app/models/exporters/solr/irobject_exporter.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class Exporters::Solr::IrobjectExporter < Exporters::Solr::BaseExporter
```
[the remark this answers is not in the document]
Comment (mbarnett) [reply app/models/exporters/solr/irobject_exporter.rb]: Nope! Older file I thought I'd deleted before I'd pushed. I'll push an update to remove it.

File: app/models/exporters/exporters.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+module Exporters
+  extend ActiveSupport::Autoload
```
[the remark this answers is not in the document]
Comment (mbarnett) [reply app/models/exporters/exporters.rb]: Whoops, also left over from an earlier version

File: app/models/exporters/solr/base_exporter.rb [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+  end
+
+  def self.solr_roles_for(name)
+    name_to_roles_map[name]
+  end
+
+  def self.facet?(name)
+    raise Exporters::Solr::UnknownAttributeError, "no such attribute #{name}" unless name_to_roles_map.key?(name)
+
+    (name_to_roles_map[name] & SOLR_FACET_ROLES).present?
```
[the remark this answers is not in the document]
Comment (mbarnett) [reply app/models/exporters/solr/base_exporter.rb]: So, concrete example: Item has a sort_year. In ItemExporter we have:

```ruby
  index :sort_year, type: :integer, role: [:search, :sort, :range_facet]
```

The `name_to_roles_map` therefore maps `:sort_year` to `[:search, :sort, :range_facet]`.

We need to be able to ask if a given attribute is a "SOLR_FACET_ROLE" so that we know when we generated searches for an Item, we can say "when you do this search, here are all of the attribute names that we want facet results for". Specifically the `:facet`, `:range_facet`, and `:pathing` roles have this automatic "ask for facet results for this index" behaviour.

So, to summarize, we have some roles for `sort_year`: `[:search, :sort, :range_facet]`. We have some roles we call SOLR_FACET_ROLES: `[:facet, :range_facet, :pathing]`

What we want to know is, are any of the roles in the former present in the latter? Put another way, we want to know if there's any _overlap_ between these two things, or to say the same thing even more mathy, we're looking to see if the intersection between these two sets of things is empty or not.

[&](https://ruby-doc.org/core-2.4.1/Array.html#method-i-26) is a method on Array returning the intersection of the array with another array. 

```ruby
[:search, :sort, :range_facet] & [:facet, :range_facet, :pathing]
=> [:range_facet]

[:search, :sort] & [:facet, :range_facet, :pathing]
=> []
```

We therefore return whether or not the intersection is `present?`, which is true if there is some overlap between arrays, and false otherwise.

File: app/models/jupiter_core/solr_services/name_mangling.rb (hunk 3 above)
Comment (mbarnett) [reply app/models/jupiter_core/solr_services/name_mangling.rb]: `singular` is a leftover from me mapping out some Solrizer behaviour to verify which index roles Solr allows only a single value for. I'll remove it.

You're completely right when you say " this could also be done by improving the dynamic type naming on the Solr side." and in a perfect world I'm 100% on board with that. When we discussed this originally we talked about either doing that or just eliminating the dynamic typing and solrizer-style name mangling entirely. When we were mapping out this work in May, though, it occurred to me that we already have bookmarks and (possibly?) some indexes out in the wild with these `title_tesim` style index names in the URLs.

The net result was that rather than look at what inbound links are in the wild, and maybe have to do some redirection to handle that, for the scope of this work we settled on leaving the Solr schema dynamic field types unchanged and having our new code map to the same dynamic stems that Solrizer would have, so that all of the URLs remain the same.

tl;dr the Solr schema dynamic type names are awful but we're preserving them for backwards compatibility at this time.

File: app/models/exporters/solr/base_exporter.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (pgwillia) [reply app/models/exporters/solr/base_exporter.rb]: Thanks.  I learned something new.  & is a method on Array for set intersection :)

**Comment by mbarnett** [comment]
Finally got around to deleting those files, so this should be good to go now.
